### ISSUE TITLE
dockerd eating too much memory and keeps growing

### ISSUE THREAD

**Issue by nasen23** [moby/moby#45842]
### Description

We have deployed kubernetes(rke) on a single node and some applications on it. Sometimes memory usage of dockerd process keeps growing and cpu usage is also very high. Finally that would take up all memory resources on that machine and we have to reboot.

### Reproduce

Don't think this could be easily reproduced. This might be related to the applications. Arbitrarily speaking the steps could be deploying a rke kubernetes cluster, running grafana's loki stack and running some applications that produce numerous logs.

### docker version

```bash
Client: Docker Engine - Community
 Version:           24.0.2
 API version:       1.43
 Go version:        go1.20.4
 Git commit:        cb74dfc
 Built:             Thu May 25 21:52:22 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.2
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.4
  Git commit:       659604f
  Built:            Thu May 25 21:52:22 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.21
  GitCommit:        3dce8eb055cbb6872793272b4f20ed16117344f8
 runc:
  Version:          1.1.7
  GitCommit:        v1.1.7-0-g860f061
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client: Docker Engine - Community
 Version:    24.0.2
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.5
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.18.1
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 486
  Running: 302
  Paused: 0
  Stopped: 184
 Images: 264
 Server Version: 24.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 nvidia runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dce8eb055cbb6872793272b4f20ed16117344f8
 runc version: v1.1.7-0-g860f061
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
 Kernel Version: 5.15.0-73-generic
 Operating System: Ubuntu 20.04.6 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 48
 Total Memory: 502.6GiB
 Name: server
 ID: R3MZ:NUMU:RKFE:VSG2:V5PB:VOFO:UB3U:7NUS:F45X:73TR:RAXZ:KUSY
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 5270
  Goroutines: 2397
  System Time: 2023-06-29T10:56:13.096261094+08:00
  EventsListeners: 0
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Registry Mirrors:
  https://hub-mirror.c.163.com/
 Live Restore Enabled: true
```

### Additional Info

I've run go pprof to capture some information about cpu usage and memory usage when dockerd memory is growing fast. You can click the two images below to see the full image.

![cpu profile](https://github.com/moby/moby/assets/52596266/0e3d8e1f-18d8-4147-87a7-32c1f9f8d5c3)
![profile002](https://github.com/moby/moby/assets/52596266/420291a1-09a9-484b-9d16-4c9a68ef025f)

Also the original pprof files are attached. 
[dockerd_prof.zip](https://github.com/moby/moby/files/11901640/dockerd_prof.zip)

The applications we're running on this kubernetes cluster are one that would create many argo workflows to do some sort of experiment and a log scraper (grafana's loki + promtail). The log volume is pretty numerous (sometimes more than 10k lines/s). Don't know if that could be responsible for this.

**Comment by thaJeztah** [comment]
<img width="1022" alt="Screenshot 2023-06-29 at 12 01 19" src="https://github.com/moby/moby/assets/1804568/5351a454-f077-4861-b7ed-e7bac9798f98">


I have not looked at the `dockerd_prof` you provided, but from that output, it looks like (relatively) a lot of time is taken up by `GetTotalUsedFds`, which in itself doesn't do "much" (it's only counting the number of directories in `/proc/self/fd`
https://github.com/moby/moby/blob/14de8ba4ead0cd414dc3bbdab2c88b97e9840bc0/pkg/fileutils/fileutils_unix.go#L15-L22

But perhaps with many file descriptors opened it may accumulate some resources in doing so. We could probably optimise that code a bit (I was looking at that earlier Today). Not sure if that's the root cause of issues, but perhaps with many tools accessing files, perhaps congestion is happening somewhere.

> The applications we're running on this kubernetes cluster are one that would create many argo workflows to do some sort of experiment and a log scraper (grafana's loki + promtail). The log volume is pretty numerous (sometimes more than 10k lines/s). Don't know if that could be responsible for this.

Is this scraping container logs? I should definitely add there that all files inside /var/lib/docker are designed to be exclusively accessed by the docker engine. While it may be _possible_ to access files out of band, doing so has been a known cause for issues (e.g. other tools trying to concurrently access log-files).

If those tools also "hammer" the `/info` endpoint (which can be a "costly" endpoint as it's collecting data from many parts), perhaps it's the combination.

**Comment by nasen23** [comment]
Thanks for the fast reply.
I do see a lot of calls to `/info` endpoint in dockerd log. And I found `cri-dockerd` is guilty for this. So it's indeed the combination. 

I have 1 question though. Those log scraping tools has to read container logs from /var/lib/docker. But how's that related to the cost in `/info` api or the function `GetTotalUsedFds`?

**Comment by corhere** [comment]
- Relevant: https://github.com/Mirantis/cri-dockerd/pull/209

-----

> I have 1 question though. Those log scraping tools has to read container logs from /var/lib/docker. But how's that related to the cost in `/info` api or the function `GetTotalUsedFds`?

Other software accessing files in /var/lib/docker out of band might have something to do with dockerd having a large number of open file descriptors, or it might be causing other unrelated issues which are contributing to the excess memory consumption.

**Comment by thaJeztah** [comment]
@corhere should we backport https://github.com/moby/moby/pull/45856 to 23.0 as well?

**Comment by corhere** [comment]
<img width="529" alt="Screenshot 2023-06-30 at 1 11 43 PM" src="https://github.com/moby/moby/assets/274527/aed660f2-abcb-43d4-8f3e-376d1c33684b">

**Comment by nasen23** [comment]
I guess this issue could be closed. I upgraded my `cri-dockerd` and this out-of-memory issue is fixed. Thank you all along.

**Comment by neersighted** [comment]
Closing as the immediate issue in moby/moby has been addressed, and cri-dockerd 0.3.4 caches calls to `GET /info` to further reduce pressure on the daemon.
